### PR TITLE
CoreFoundation: apply hook to correct platform

### PR DESCRIPTION
**Pull request by szlend**
## Description of changes

Cross-compiling from darwin to linux is broken if the derivation includes CoreFoundation somewhere in its (native) input deps. This change makes the cf-setup-hook aware of its platform role, replacing NIX_LDFLAGS with NIX_LDFLAGS_FOR_BUILD when necessary.

Fixes https://github.com/NixOS/nixpkgs/issues/278348 in `23.11` without getting rid of the hook in case anyone was relying on it.

In `unstable` we probably want to get rid of the hook entirely. See #284629.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by szlend** [comment]
Some checks failed because of API rate-limiting. I don't think I have permissions to rerun them.

**Review by reckenrode — APPROVED** [review]
LGTM

@toonn Pinging you for your review.

File: pkgs/os-specific/darwin/apple-sdk-11.0/cf-setup-hook.sh [hunk 1]
```diff
@@ -1,6 +1,14 @@
...
+  local role_post
+  getHostRole
+
+  local cflags_compile_role_var="NIX_CFLAGS_COMPILE${role_post}"
+  local cflags_compile_role="${!cflags_compile_role_var-}"
+  export NIX_CFLAGS_COMPILE${role_post}="-F@out@/Library/Frameworks${cflags_compile_role:+ $cflags_compile_role}"
+
+  local ldflags_role_var="NIX_LDFLAGS${role_post}"
+  local ldflags_role="${!ldflags_role_var-}"
+  export NIX_LDFLAGS${role_post}+="${ldflags_role:+ }@out@/Library/Frameworks/CoreFoundation.framework/CoreFoundation.tbd"
```
Comment (toonn) [inline pkgs/os-specific/darwin/apple-sdk-11.0/cf-setup-hook.sh:11]: Isn't this appending the entire value of `NIX_LDFLAGS${role_post}` again? Do we want to be outright setting the variable here or keep it to just appending the CoreFoundation.tbd path?

File: pkgs/os-specific/darwin/apple-sdk-11.0/cf-setup-hook.sh [hunk 2]
```diff
@@ -1,6 +1,14 @@
...
 }
 
-preConfigureHooks+=(forceLinkCoreFoundationFramework)
```
Comment (toonn) [inline pkgs/os-specific/darwin/apple-sdk-11.0/cf-setup-hook.sh:6]: What's the motivation for this change? Not saying it's wrong, just not understanding it.

File: pkgs/os-specific/darwin/apple-sdk-11.0/cf-setup-hook.sh [hunk 3]
```diff
@@ -1,6 +1,14 @@
 forceLinkCoreFoundationFramework() {
-  NIX_CFLAGS_COMPILE="-F@out@/Library/Frameworks${NIX_CFLAGS_COMPILE:+ }${NIX_CFLAGS_COMPILE-}"
```
Comment (toonn) [inline pkgs/os-specific/darwin/apple-sdk-11.0/cf-setup-hook.sh:2]: Do we actually want to include `-F@out@/Library/Frameworks` if `NIX_CFLAGS_COMPILE` (suffixed or not) is unset or empty?

It's valid if we just want to maintain this behavior unchanged but I do wonder whether it makes sense.

**Review by toonn — COMMENTED** [review]
Couple questions that apply to the hook changes, only commented on one of them because the changes are identical AFAICT.

File: pkgs/os-specific/darwin/apple-sdk-11.0/cf-setup-hook.sh (hunk 2 above)
Comment (szlend) [reply pkgs/os-specific/darwin/apple-sdk-11.0/cf-setup-hook.sh:6]: `preConfigureHooks` move the execution of the function in another scope. `getHostRole` and as a result `role_post` are only available in this scope.

File: pkgs/os-specific/darwin/apple-sdk-11.0/cf-setup-hook.sh (hunk 1 above)
Comment (szlend) [reply pkgs/os-specific/darwin/apple-sdk-11.0/cf-setup-hook.sh:11]: `${ldflags_role:+ }` prepends a space if the variable is not empty. It doesn't include the content of the variable. It's probably not necessary and we could always include an extra space to simplify it. I just tried to keep the functionality as close as possible to the original.

File: pkgs/os-specific/darwin/apple-sdk-11.0/cf-setup-hook.sh (hunk 3 above)
Comment (szlend) [reply pkgs/os-specific/darwin/apple-sdk-11.0/cf-setup-hook.sh:2]: Hm, I'm not sure I follow. Is it safe to assume that `NIX_CFLAGS_COMPILE*` is always set? Because if it's not then this is definitely necessary. Or is the question whether `-F@out@/Library/Frameworks` is necessary in the first place?

Either way, this hook is most likely getting removed in `24.05`, so I wouldn't worry too much about it.

File: pkgs/os-specific/darwin/apple-sdk-11.0/cf-setup-hook.sh (hunk 2 above)
Comment (toonn) [reply pkgs/os-specific/darwin/apple-sdk-11.0/cf-setup-hook.sh:6]: When does this run when not explicitly added to one of the *Hooks variables?

File: pkgs/os-specific/darwin/apple-sdk-11.0/cf-setup-hook.sh (hunk 3 above)
Comment (toonn) [reply pkgs/os-specific/darwin/apple-sdk-11.0/cf-setup-hook.sh:2]: The latter is what I meant. Like I said, I'm not really worried about it. Just jumps out as a bit weird.

File: pkgs/os-specific/darwin/apple-sdk-11.0/cf-setup-hook.sh (hunk 1 above)
Comment (toonn) [reply pkgs/os-specific/darwin/apple-sdk-11.0/cf-setup-hook.sh:11]: Totally right, misread the expansion entirely. That particular pattern is very confusing in a Github code block that wraps at whitespace >.<

File: pkgs/os-specific/darwin/apple-sdk-11.0/cf-setup-hook.sh (hunk 2 above)
Comment (szlend) [reply pkgs/os-specific/darwin/apple-sdk-11.0/cf-setup-hook.sh:6]: Actually, my description was slightly incorrect. Here's where it's executed:

https://github.com/NixOS/nixpkgs/blob/fec2bb16200119774e6f34e5d1d65d3a5eb7c61f/pkgs/stdenv/generic/setup.sh#L669-L671

`getHostRole` requires `$hostOffset` which is not available in the configuration phase. It needs to run in the scope of `activatePackage`.

Comment (szlend) [reply pkgs/os-specific/darwin/apple-sdk-11.0/cf-setup-hook.sh:6]: `activatePackage` runs for every build input in your derivation. Its purpose is to configure `PATH` and run setup hooks in the context of your derivation. This means the setup hook is aware of how its being included in the derivation (either through `buildInputs` or `nativeBuildInput`). That context is lost when you postpone the execution of the hook to `preConfigure`.

I'm not sure why this was originally configured in the `preConfigureHooks`. But I think configuring compiler/linker flags here should be fine. That's what setup hooks are for, no?

**Review by toonn — APPROVED** [review]
LGTM
